include/ttmlir/Dialect/TTNN/Analysis/DisjointL1ChainConfigsUnion.h

```diff
@@ -0,0 +1,56 @@
+// SPDX-FileCopyrightText: (c) 2024 Tenstorrent AI ULC
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef TTMLIR_DIALECT_TTNN_ANALYSIS_DISJOINTL1CHAINCONFIGSUNION_H
+#define TTMLIR_DIALECT_TTNN_ANALYSIS_DISJOINTL1CHAINCONFIGSUNION_H
+
+#include "ttmlir/Dialect/TTNN/Analysis/L1ChainConfig.h"
+#include <llvm/ADT/SmallVector.h>
+#include <mlir/Dialect/Func/IR/FuncOps.h>
+#include <mlir/IR/Operation.h>
+
+namespace mlir::tt::ttnn {
+
+class DisjoinL1ChainConfigsUnion {
+private:
+  llvm::DenseMap<Operation *, L1ChainConfig> l1ChainConfigsMap;
+  llvm::DenseMap<Operation *, Operation *> parents;
+
+public:
+  DisjoinL1ChainConfigsUnion() = default;
+
+  // Inserts OpL1MemSpec in the same L1ChainConfig that op belongs to.
+  // In case the op is not provided, a new L1ChainConfig is created.
+  void insertOpL1MemSpec(OpL1MemSpec opL1MemSpec, Operation *referenceOp);
+
+  // Inserts new L1ChainConfig in the union and construct
+  // the parent tree for its ops.
+  void insertL1ChainConfig(L1ChainConfig &l1ChainConfig);
+
+  /** @return the "representative" op in op's component */
```

Comment:
Based on the representative op we can fetch the adequate `L1ChainConfig` since the `l1ChainConfigsMap` has only the representative ops as values. Basically, the size of `l1ChainConfigsMap` is equal to the number of l1ChainConfigs and not the number of ops.